TRUNCATE PARTITION ALL gets stuck, ddl_worker retries, error returned

## Bug Report


### 1. Minimal reproduce step (Required)

```sql
create table t (a int primary key) partition by hash (a);
alter table t truncate partition all;
```

### 2. What did you expect to see? (Required)

Speed up execution and return _understandable_ error messages. Now it runs for several minutes (5min+) and fails.

### 3. What did you see instead (Required)

````
[2021/02/03 10:27:26.695 +00:00] [INFO] [error.go:25] ["ERROR 1735: DDL JOB ROLLBACK, ERROR MSG: UNKNOWN PARTITION 'DROP?' IN TABLE 'TABLE_5
_BINARY_2'"]
````

You should be able to see log like the following:

````
[2021/02/03 11:32:00.966 +00:00] [ERROR] [ddl_worker.go:720] ["[ddl] run DDL job error"] [worker="worker 3, tp general"] [error="[table:1735]Unknown partition 'drop?' in table 'table_5_binary_4'"]
[2021/02/03 11:32:00.975 +00:00] [INFO] [ddl_worker.go:499] ["[ddl] run DDL job failed, sleeps a while then retries it."] [worker="worker 3, tp general"] [waitTime=1s] [error="[table:1735]Unknown partition 'drop?' in table 'table_5_binary_4'"]
[2021/02/03 11:32:01.975 +00:00] [INFO] [ddl_worker.go:774] ["[ddl] schema version doesn't change"] [worker="worker 3, tp general"]
[2021/02/03 11:32:01.976 +00:00] [INFO] [ddl_worker.go:599] ["[ddl] run DDL job"] [worker="worker 3, tp general"] [job="ID:1198, Type:truncate partition, State:running, SchemaState:none, SchemaID:1190, TableID:1192, RowCount:0, ArgLen:0, start time: 2021-02-03 11:31:09.291 +0000 UTC, Err:[table:1735]Unknown partition 'drop?' in table 'table_5_binary_4', ErrCount:52, SnapshotVersion:0"]
[2021/02/03 11:32:01.978 +00:00] [ERROR] [ddl_worker.go:720] ["[ddl] run DDL job error"] [worker="worker 3, tp general"] [error="[table:1735]Unknown partition 'drop?' in table 'table_5_binary_4'"]
[2021/02/03 11:32:01.986 +00:00] [INFO] [ddl_worker.go:499] ["[ddl] run DDL job failed, sleeps a while then retries it."] [worker="worker 3, tp general"] [waitTime=1s] [error="[table:1735]Unknown partition 'drop?' in table 'table_5_binary_4'"]
````

### 4. What is your TiDB version? (Required)

````
-----------------------------------+
| Release Version: v5.0.0-rc
Edition: Community
Git Commit Hash: fd4437dde445012f5b75e5d4f336433cb076c1ee
Git Branch: heads/refs/tags/v5.0.0-rc
UTC Build Time: 2021-01-12 06:00:03
GoVersion: go1.13
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
````

--- a/ddl/db_test.go
+++ b/ddl/db_test.go
@@ -6588,6 +6588,7 @@ func (s *testSerialDBSuite) TestIssue22819(c *C) {
 	tk1.MustExec("update t1 set v = 2 where v = 1")
 
 	tk2.MustExec("alter table t1 truncate partition p0")
+	tk2.MustExec("alter table t1 truncate partition all")
 
 	_, err := tk1.Exec("commit")
 	c.Assert(err, ErrorMatches, ".*8028.*Information schema is changed during the execution of the statement.*")

--- a/ddl/db_test.go
+++ b/ddl/db_test.go
@@ -6588,8 +6588,20 @@ func (s *testSerialDBSuite) TestIssue22819(c *C) {
 	tk1.MustExec("update t1 set v = 2 where v = 1")
 
 	tk2.MustExec("alter table t1 truncate partition p0")
-	tk2.MustExec("alter table t1 truncate partition all")
 
 	_, err := tk1.Exec("commit")
 	c.Assert(err, ErrorMatches, ".*8028.*Information schema is changed during the execution of the statement.*")
+}
+
+func (s *testSerialSuite) TestTruncateAllPartitions(c *C) {
+	tk1 := testkit.NewTestKit(c, s.store)
+	tk1.MustExec("use test;")
+	tk1.MustExec("drop table if exists partition_table;")
+	defer func() {
+		tk1.MustExec("drop table if exists partition_table;")
+	}()
+	tk1.MustExec("create table partition_table (v int) partition by hash (v) partitions 2;")
+	tk1.MustExec("insert into partition_table values (1);")
+	tk1.MustExec("alter table partition_table truncate partition all;")
+	tk1.MustQuery("select count(*) from partition_table;").Check(testkit.Rows("0"))
 }

--- a/ddl/db_test.go
+++ b/ddl/db_test.go
@@ -6600,8 +6600,8 @@ func (s *testSerialSuite) TestTruncateAllPartitions(c *C) {
 	defer func() {
 		tk1.MustExec("drop table if exists partition_table;")
 	}()
-	tk1.MustExec("create table partition_table (v int) partition by hash (v) partitions 2;")
-	tk1.MustExec("insert into partition_table values (1);")
+	tk1.MustExec("create table partition_table (v int) partition by hash (v) partitions 10;")
+	tk1.MustExec("insert into partition_table values (0),(1),(2),(3),(4),(5),(6),(7),(8),(9),(10);")
 	tk1.MustExec("alter table partition_table truncate partition all;")
 	tk1.MustQuery("select count(*) from partition_table;").Check(testkit.Rows("0"))
 }

--- a/ddl/ddl_api.go
+++ b/ddl/ddl_api.go
@@ -2971,12 +2971,19 @@ func (d *ddl) TruncateTablePartition(ctx sessionctx.Context, ident ast.Ident, sp
 	}
 
 	pids := make([]int64, len(spec.PartitionNames))
-	for i, name := range spec.PartitionNames {
-		pid, err := tables.FindPartitionByName(meta, name.L)
-		if err != nil {
-			return errors.Trace(err)
+	if spec.OnAllPartitions {
+		pids = make([]int64, len(meta.GetPartitionInfo().Definitions))
+		for i, def := range meta.GetPartitionInfo().Definitions {
+			pids[i] = def.ID
 		}
-		pids[i] = pid
+	} else {
+		for i, name := range spec.PartitionNames {
+			pid, err := tables.FindPartitionByName(meta, name.L)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			pids[i] = pid
+		}
 	}
 
 	job := &model.Job{

--- a/go.mod
+++ b/go.mod
@@ -92,6 +92,4 @@ require (
 	sourcegraph.com/sourcegraph/appdash-data v0.0.0-20151005221446-73f23eafcf67
 )
 
-replace github.com/pingcap/parser v0.0.0-20210310110710-c7333a4927e6 => github.com/AndrewDi/parser v0.0.0-20210313124806-d0cf71c13399
-
 go 1.13

--- a/go.mod
+++ b/go.mod
@@ -50,7 +50,7 @@ require (
 	github.com/pingcap/goleveldb v0.0.0-20191226122134-f82aafb29989
 	github.com/pingcap/kvproto v0.0.0-20210308063835-39b884695fb8
 	github.com/pingcap/log v0.0.0-20201112100606-8f1e84a3abc8
-	github.com/pingcap/parser v0.0.0-20210311132237-9841cb715606
+	github.com/pingcap/parser v0.0.0-20210314080929-ed8900c94180
 	github.com/pingcap/sysutil v0.0.0-20210221112134-a07bda3bde99
 	github.com/pingcap/tidb-tools v4.0.9-0.20201127090955-2707c97b3853+incompatible
 	github.com/pingcap/tipb v0.0.0-20210309080453-72c4feaa6da7

--- a/go.mod
+++ b/go.mod
@@ -91,5 +91,7 @@ require (
 	sourcegraph.com/sourcegraph/appdash v0.0.0-20190731080439-ebfcffb1b5c0
 	sourcegraph.com/sourcegraph/appdash-data v0.0.0-20151005221446-73f23eafcf67
 )
-replace github.com/pingcap/parser v0.0.0-20210310110710-c7333a4927e6 => github.com/AndrewDi/parser v0.0.0-20210301125952-abc04df118c2
+
+replace github.com/pingcap/parser v0.0.0-20210310110710-c7333a4927e6 => github.com/AndrewDi/parser v0.0.0-20210313124806-d0cf71c13399
+
 go 1.13

--- a/go.mod
+++ b/go.mod
@@ -91,5 +91,5 @@ require (
 	sourcegraph.com/sourcegraph/appdash v0.0.0-20190731080439-ebfcffb1b5c0
 	sourcegraph.com/sourcegraph/appdash-data v0.0.0-20151005221446-73f23eafcf67
 )
-
+replace github.com/pingcap/parser v0.0.0-20210310110710-c7333a4927e6 => github.com/AndrewDi/parser v0.0.0-20210301125952-abc04df118c2
 go 1.13

